### ISSUE TITLE
YCSB error 'Signature not yet current' when authentication enabled 

### ISSUE THREAD

**Issue by fruch** [scylladb/scylla#6345]
*Installation details*
Scylla version (or git commit hash): 4.0.rc2-0.20200427.602ed43ac7a
Cluster size: 4
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0fcdfef1a8f37ec6d (eu-west-1)

## Summary
We are got this error from YCSB (aws-sdk-java 1.10.48) while using `alternator_enforce_authorization:true`

```
18730030 [Thread-6] ERROR site.ycsb.db.DynamoDBClient  -com.amazonaws.AmazonServiceException: Signature not yet current: 20200428T222206Z is still later than 20200428T220706Z (current time + 15 min.) (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: InvalidSignatureException; Request ID: null)
```

It only happen once in a 24h run.

## Logs
* Jenkins job - https://jenkins.scylladb.com/view/scylla-4.0/job/scylla-4.0/job/longevity/job/longevity-alternator-200gb-48h-test/2
* **db-cluster** - https://cloudius-jenkins-test.s3.amazonaws.com/7775b469-ab71-465b-a2d3-63ad35772cae/20200429_193841/db-cluster-7775b469.zip
* **loader-set** - https://cloudius-jenkins-test.s3.amazonaws.com/7775b469-ab71-465b-a2d3-63ad35772cae/20200429_193841/loader-set-7775b469.zip
* **monitor-set** - https://cloudius-jenkins-test.s3.amazonaws.com/7775b469-ab71-465b-a2d3-63ad35772cae/20200429_193841/monitor-set-7775b469.zip
* **sct-runner** - https://cloudius-jenkins-test.s3.amazonaws.com/7775b469-ab71-465b-a2d3-63ad35772cae/20200429_193841/sct-runner-7775b469.zip
* download monitor data - `hydra investigate show-monitor 7775b469-ab71-465b-a2d3-63ad35772cae`

**Comment by nyh** [comment]
This is very odd... The message suggests that signature_date exactly 15 minutes into the future from validity_str, which itself is 15 minutes into the future from now. What could possible cause this to occur, how is it **exactly** 15 minutes after the 15 minute we allow, and how did it happen only once in 24 hours? That's very strange. @psarna do you have any guesses?

**Comment by fruch** [comment]
I see more occurrences of it,

I think something is fishy in this:
```
std::string expiration_str = format_time_point(db_clock::now() - 15min);
```

cause the `signature_date` seem to match to time this print occurs

**Comment by psarna** [comment]
@fruch I didn't find any more occurences of `Signature not yet current`, but of the similar kind, `Signature expired`:

```
Signature expired: 20200429T132213Z is now earlier than 20200429T133714Z (current time - 15 min.) (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: InvalidSignatureException; Reque      st ID: null)
```

Now, this second one makes more sense, as it simply means that the request was prepared more than 15 minutes before it reached alternator, in which case it's considered stale and it's refused. I'm not sure how this YCSB workload works, but it looks like it prepared some statements at 13:32:13 and then they reached alternator at 13:37:14, which is right after the deadline. Perhaps YCSB, via its DynamoDB driver, has some requests already created and reuses them in the workload until they hit an error (in this case, that the signature is no longer valid)?

As for the original issue, which is `signature not yet current`, that one is extremely weird... I took the original entry from the log, since it contains timestamps:
```json
{
	"timestamp": 1588112526.5605857,
	"severity": "ERROR",
	"type": "error",
	"node": "Node alternator-48h-4-0-loader-node-7775b469-3 [18.203.102.159 | 10.0.154.205] (seed: False)",
	"stress_cmd": "bin/ycsb run dynamodb -P workloads/workloadc -threads 20 -p readproportion=0.5 -p updatepropo     rtion=0.5 -p recordcount=200200300 -p fieldcount=8 -p fieldlength=128 -p operationcount=2140000000 -p dataintegrity=true -s -P /tmp/dynamodb.properties -p maxexecutiontime=173400",
	"log_file_name": null,
	"errors": ["18730030 [Thread-6] ERROR site.ycsb.db.DynamoDBClient  -com.amazonaw     s.AmazonServiceException: Signature not yet current: 20200428T222206Z is still later than 20200428T220706Z (current time + 15 min.) (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: InvalidSignatureException; Request ID: null)"]
}
```
and it's worrysome, since the timestamp translates to `2020-04-28T22:22:06+00:00`. It means that YCSB did not try to send any custom time - it simply sent its local time, since the log entry has exactly the same timestamp as in the signature sent by the client. That in turn means that the alternator server, a.k.a. Scylla, deeply believes that current time is actually `21:52` now, which is precisely half an hour early.

edit: of course... took me a while to remember that `std::gmtime` used in alternator is one of these really treacherous functions that use static buffers inside them, so they're totally not realiable for threads... It simply *must* be the fact that due to a race, expiration time can get overwritten by/with validity, in which case the values we compare are not the ones we want to compare with.

The solution is simple - use `gmtime_r`, which is a safe vartiant of `gmtime`. By the way, I also see that we use raw, unsafe `gmtime` in `clocks-impl.cc` when formatting timestamps - we should switch to the safe variant in there as well.

**Comment by fruch** [comment]
yep this sound more like it, since it was too cleanly match the other things we compute at the same time...

and I see the patches are on route :) 
nice catch @psarna, I was staring at that code for a while, trying to see if I can spot a clear issue, didn't got to dig that further.

**Comment by nyh** [comment]
Backported the bugfix to 4.0 (commit abd73cab78114c62ac132a83c7ff8d1e73132b8b) and and 3.3 (commit 3216a1a70aa3d45a496e08d01a26d04de89ace88).

**Comment by nyh** [comment]
The "backport candidate" label was re-added by the bot by mistake: The fix is already in 4.1, and backported to 4.0 and 3.3. So removing the label, again.
